### PR TITLE
KNIP-894: Replacement of nodes Proactive

### PR DESCRIPTION
**Pull request by prsurve**
Signed-off-by: Pratik Surve <pratiksurve2130@gmail.com>

### CODE REVIEW

File: ocs_ci/ocs/node.py [hunk 1]
```diff
@@ -359,3 +359,22 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
         }
 
     return utilization_dict
+
+
+def get_node_from_machine_name(machine_name):
+    """
+    Get node name from a give machine_name
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: given

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from ocs_ci.ocs import constants
+from ocs_ci.ocs import ocp
+from ocs_ci.ocs import node
+from tests.sanity_helpers import Sanity
+
+log = logging.getLogger(__name__)
+
+
+@tier4
+@tier4b
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: please change to tier4a as this tier is being executed more frequently and we'd like these tests to be there

File: ocs_ci/ocs/node.py [hunk 3]
```diff
@@ -359,3 +359,22 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+
+    node_name (str): Name of Machine
+
+    Returns:
+        obj: Object of node
+    """
+    node_list = get_node_objs()
+    matching_annotation = "openshift-machine-api/" + machine_name
+    for nodes in node_list:
```
Comment (RazTamir) [inline ocs_ci/ocs/node.py]: for *node* (instead of nodes)

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+from ocs_ci.framework.pytest_customization.marks import tier4b
+from ocs_ci.ocs.resources import pod
+
+from ocs_ci.framework.testlib import (
+    tier4, ManageTest, aws_platform_required, ignore_leftovers, ipi_deployment_required)
+from ocs_ci.ocs import machine
+from ocs_ci.ocs import constants
+from ocs_ci.ocs import ocp
+from ocs_ci.ocs import node
```
Comment (RazTamir) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: please import all in 1 line:
from ocs_ci_ocs import (
    machine, constants, ocp, node
)

Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: ocs_ci/ocs/node.py (hunk 3 above)
Comment (prsurve) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 1 above)
Comment (prsurve) [reply ocs_ci/ocs/node.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 2 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: ocs_ci/ocs/node.py [hunk 5]
```diff
@@ -359,3 +359,22 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
         }
 
     return utilization_dict
+
+
+def get_node_from_machine_name(machine_name):
```
Comment (PrasadDesala) [inline ocs_ci/ocs/node.py]: Instead, you can get the machine obj and get the node name using it  as below,

machine_objs = get_machine_objs()
    for machine in machine_objs:
        if machine.get().get(
                'status'
        ).get('addresses')[1].get('address')

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
+import logging
+import re
+
+import pytest
+import random
+
+from tests.helpers import get_worker_nodes
+
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Remove new line space

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+from ocs_ci.framework.pytest_customization.marks import tier4a
+from ocs_ci.ocs.resources import pod
+
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Remove new line space

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        self.sanity_helpers = Sanity()
+
+    def test_nodereplacement_proactive(self, pvc_factory, pod_factory, dc_pod_factory):
+        """
+        Knip-894 Node Replacement proactive
+
+        """
+
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Remove new line space

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        log.info("Creating dc pod backed with cephfs pvc and running io in bg")
+        for worker_node in worker_node_list:
+            if worker_node != osd_node_name:
+                cephfs_dc_pod = dc_pod_factory(interface=constants.CEPHFILESYSTEM, node_name=worker_node, size=20)
+                pod.run_io_in_bg(cephfs_dc_pod, expect_to_fail=False, fedora_dc=True)
+
+        # Unscheduling node
+        node.unschedule_nodes([osd_node_name])
+        log.info("Getting machine name from specified node name")
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: After unscheduling the node, we have to drain the node as apart of procedure.

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        machine_name = machine.get_machine_from_node_name(osd_node_name)
+        log.info(f"Node {osd_node_name} associated machine is {machine_name}")
+        log.info(f"Deleting machine {machine_name}")
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: We are also waiting for the new machine to come up. This log message has to be updated with that info as well

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        machine.delete_machine_and_check_state_of_new_spinned_machine(machine_name)
+        new_machine_list = machine.get_machines()
+        for machines in new_machine_list:
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Is this for getting the new spun node name?
We can get the worker list before machine delete and after machine delete and can find out the new spun machine.

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            # Trimming is done to get just machine name
+            # eg:- machine_name:- prsurve-40-ocs-43-kbrvf-worker-us-east-2b-nlgkr
+            # After trimming:- prsurve-40-ocs-43-kbrvf-worker-us-east-2b
+            if re.match(machines.name[:-6], machine_name):
+                new_machine_name = machines.name
+        machineset_name = machine.get_machineset_from_machine_name(new_machine_name)
+        log.info("Wating for new worker node to be in ready state")
+        machine.wait_for_new_node_to_be_ready(machineset_name)
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Isn't delete_machine_and_check_state_of_new_spinned_machine() waits till the new machine reaches running state?
If so, then this is not needed as any how we are checking the state of the machine above

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        log.info(
+            f"Successfully labeled {new_node_obj.name} with OCS storage label"
+        )
+        # Creating Resources
+        log.info("Creating Resources using sanity helpers")
+        self.sanity_helpers.create_resources(pvc_factory, pod_factory)
+        # Deleting Resources
+        self.sanity_helpers.delete_resources()
+        # Verify everything running fine
+        log.info("Verifying All resources are Running and matches expected result")
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: We will also have to verify below,
*. Create DC app pods on the failure node and start IO before inducing failure.
*. Post failure, the DC app pods should automatically get created on the other nodes 
*. We should be able to start IO on the DC app pods.
FYI, I have a patch for doing above which is still under review, https://github.com/red-hat-storage/ocs-ci/pull/1591
I am okay if you have plans to cover this in another PR :)

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 12 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: No it does not

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 13 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: will cover in next pr when your pr gets merged

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 11 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: this function is used to get new machine name after old is deleted

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 8 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 9 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 10 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 7 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 6 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: ocs_ci/ocs/node.py (hunk 5 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: @prsurve This comment is not addressed. Any particular reason?

Comment (prsurve) [reply ocs_ci/ocs/node.py]: i have made changes local i need to test and will update shortly

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 11 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: How this functions deals incase of machinset having more than one machines in it?

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    @pytest.fixture(autouse=True)
+    def init_sanity(self):
+        """
+        Initialize Sanity instance
+
+        """
+        self.sanity_helpers = Sanity()
+
+    def test_nodereplacement_proactive(self, pvc_factory, pod_factory, dc_pod_factory):
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: @PrasadDesala there is some duplicity with the test function from PR https://github.com/red-hat-storage/ocs-ci/pull/1491.

Please try avoiding this. An option would be to create a common function that will be called from both test methods

Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: @ebenahar 
@prsurve is working on this proactive pacth. 
I have submitted for Reactive PR #1607
So, you are suggesting us to cover both Proactive and reactive node replacement in a single test.py file?

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 11 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: for each worker, master there is separate machine created

File: ocs_ci/ocs/node.py (hunk 5 above)
Comment (prsurve) [reply ocs_ci/ocs/node.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+log = logging.getLogger(__name__)
+
+
+@tier4
+@tier4a
+@ignore_leftovers
+@aws_platform_required
+@ipi_deployment_required
+class TestNodeReplacement(ManageTest):
```
Comment (RazTamir) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Code looks good but I don't see even one assert statement so I wonder where are we verifying things are working as expected

Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: All the function used are having the assert added so i feel no any addition assert is required

**Pull request merged**